File: Easy/599_Minimum_Index_Sum_of_Two_Lists/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        Map<String, Integer> mp = new HashMap<>();

        for (int i = 0; i < list1.length; i++) {
            mp.put(list1[i], i);
        }

        int mn = Integer.MAX_VALUE;
        List<String> ans = new ArrayList<>();

        for (int i = 0; i < list2.length; i++) {
            if (mp.containsKey(list2[i])) {
                int sum = mp.get(list2[i]) + i;

                if (sum < mn) {
                    mn = sum;
                    ans.clear();
                    ans.add(list2[i]);
                }
                else if (sum == mn) {
                    ans.add(list2[i]);
                }
            }
        }

        return ans.toArray(new String[0]);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        String[] list11 = {"Shogun","Tapioca Express","Burger King","KFC"};
        String[] list12 = {"Piatti","The Grill at Torrey Pines","Hungry Hunter Steakhouse","Shogun"};
        String[] ans1 = solution.findRestaurant(list11, list12);
        System.out.println(ans1.toString());

        // test cases 2
        String[] list21 = {"Shogun","Tapioca Express","Burger King","KFC"};
        String[] list22 = {"KFC","Shogun","Burger King"};
        String[] ans2 = solution.findRestaurant(list21, list22);
        System.out.println(ans2.toString());

        // test cases 3
        String[] list31 = {"happy","sad","good"};
        String[] list32 = {"sad","happy","good"};
        String[] ans3 = solution.findRestaurant(list31, list32);
        System.out.println(ans3.toString());
        
    }
}

File: Easy/599_Minimum_Index_Sum_of_Two_Lists/Solution.py
from typing import List, Optional


class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        dct = dict()

        for i in range(len(list1)):
            dct[list1[i]] = i

        mn = len(list1) + len(list2)
        ans = []

        for i in range(len(list2)):
            if list2[i] in dct:
                index_sum = dct[list2[i]] + i

                if index_sum < mn:
                    mn = index_sum
                    ans.clear()
                    ans.append(list2[i])
                elif index_sum == mn:
                    ans.append(list2[i])
                
        return ans


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    list11 = ["Shogun","Tapioca Express","Burger King","KFC"]
    list12 = ["Piatti","The Grill at Torrey Pines","Hungry Hunter Steakhouse","Shogun"]
    ans1 = solution.findRestaurant(list11, list12)
    print(ans1)

    # test cases 2
    list21 = ["Shogun","Tapioca Express","Burger King","KFC"]
    list22 = ["KFC","Shogun","Burger King"]
    ans2 = solution.findRestaurant(list21, list22)
    print(ans2)

    # test cases 3
    list31 = ["happy","sad","good"]
    list32 = ["sad","happy","good"]
    ans3 = solution.findRestaurant(list31, list32)
    print(ans3)
    
    